### ISSUE TITLE
Create candidate specification for a generic follow-up queries

### ISSUE THREAD

**Issue by JRegimbal** [Shared-Reality-Lab/IMAGE-server#862]
Cases exist where additional information in a rendering is desired by users and is infeasible to produce in the initial response. This additional information will be in reference to a recently made request and may be specified in different formats (e.g., a bounding box, text, etc.). How this new information will interact with the initial response is currently undefined.

* [ ] Determine expected interactions of the new information with the original rendering, and other requirements from a user perspective.
* [ ] Propose a basic overview, for the client, on how such functionality would be implemented, respecting feature requirements.
* [ ] Propose a basic overview, for the server, on how such functionality would be implemented, respecting feature requirements.
* [ ] Request comments from IMAGE team members, revising as necessary.

**Comment by JRegimbal** [comment]
After meeting with @VenissaCarolQuadros @Cybernide @AndyBaiMQC we decided that follow-up requests will likely be fully independent from one another, including the source graphic, data representing the follow-up query (likely in the form of text), and an optional referenced region of the source graphic. At this time, it is *not* anticipated that state will need to be stored on the server. Information on the *renderings* will also *not* be sent.

This is preliminary, and will be developed further when @Cybernide and @VenissaCarolQuadros produce some stories from a user perspective focused on:

* a situation where a user requests follow-up information once related to an entire graphic,
* a situation where a user requests follow-up information multiple times related to a graphic, where requests are thematically related, and
* a situation where a user requests follow-up information once related to a specific part of the graphic.

**Comment by JRegimbal** [comment]
After meeting with @jeffbl, @Cybernide, @VenissaCarolQuadros, and @AndyBaiMQC, we agreed on the following requirements:

* The query MUST contain the original graphic and the user-provided query in the form of text.
* The query MAY indicate a point or region of the graphic that was referenced by the user at the time of the query, to serve as a rough deitic gesture.
* The query SHOULD include information on the previous set of questions and responses, if any, if context is required in crafting a response.
* There MAY be a mechanism for switching between multiple LLMs or other means of producing a response to the question.
* The server MUST produce a list of renderings encoding the response, which MAY be of different types.
* The rendering types for follow-up queries SHOULD be the same as those used for initial renderings, if possible.
* The client SHOULD display these renderings differently based on the context in which the follow-up query was made.
* The client MUST track a tree structure of sources and responses, with the original graphic and rendering as the root. Its children are the follow-up queries and associated responses. There MAY be multiple responses retained for one query, e.g., at different levels of detail.
* The server SHOULD NOT provide follow-up queries that duplicate large parts of the content in the previous rendering. That is, follow-up responses generated from a tactile graphic should not include the original graphic, with a new or modified label.

Kindly ping if anything is missing or incorrect!

**Comment by JRegimbal** [comment]
# Draft 2024-08-19

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html).

## Data Model Changes

* IMAGE requests MAY include the new top-level fields on all media types:
	* `followup`  an object with the following fields:
	   * `query` a REQUIRED field that is the question posed by the user as plain text.
	   * `focus` an OPTIONAL field that if present specifies a region of the graphic indicated by the user as `[ ulx, uly, lrx, lry ]` where the values are floating point values between 0 and 1 indicating the upper left x and y, and lower right x and y coordinates respectively.
	   * `previous` an OPTIONAL field whose value is a list of 2-length arrays. The first element of this array is a string representing the query for a request, and the second element is a string or object containing the corresponding response's data.
   * `route` a RECOMMENDED field that encodes the set of preprocessors and handlers to run with the query. It may be used independently of `followup`. It is assumed to have a value of `default` if unspecified. When present, only components that include the value of this parameter in the `ca.mcgill.a11y.image.route` list will be run.

## Server Requirements

* The orchestrator MUST check if the `route` parameter is available.
	* If it is specified and has a value that is not `default`, the orchestrator will filter preprocessors and handlers only for those that have a `ca.mcgill.a11y.image.route` label that contains the value.
	* If it is not specified or is explicitly passed `default`, then the orchestrator MUST filter for preprocessors and handlers that specify `default` or have not set the `ca.mcgill.a11y.image.route` label.
* Valid routes for  `ca.mcgill.a11y.image.route` MUST contain alphanumeric characters and underscores and are comma delimited. E.g., `"default,followup"` is valid, but `"default, follow-up queries"` is not.
* Preprocessors and handlers for follow-up queries SHOULD use existing data structures when available.
* New data structures MUST be specified using a schema and with an assigned reverse domain name identifier in the same way as existing preprocessor and renderer data structures.
* Follow-up responses SHOULD NOT duplicate large parts of the context from previous renderings in responding to the user question.
* Server responses to follow-up query requests SHALL follow the format already in use for both `/render` and `/render/preprocess` endpoints.

## Client Requirements

Functionality to send and process follow-up queries is OPTIONAL. The following specifications apply only to clients that adopt this functionality:

* Clients SHOULD allow the user to input their query in the form of language.
* The language of the query MUST match the language code contained in the request.
* Clients SHOULD display the results of follow-up requests in a way that avoids unnecessary context changes for the user.
	* Due to this, clients MAY specify different capabilities and renderers in follow-up requests than in original requests. 
	* Clients MAY also display follow-up responses in different ways than original responses of the same rendering type.
* Clients SHOULD be capable of handling multiple renderings in the response for each follow-up query.
* Clients SHOULD record the history of requests and responses, and allow the user to navigate this structure.

## Example

A client shows a tactile SVG of a dog with a frisbee to a user, who taps the frisbee and asks "what color is this?" The client sends a follow-up request that includes the following snippet:
```json
{
	/* omitted for brevity */
	"followup":{
		"query": "What color is this?",
		"focus": [0.25, 0.25, 0.30, 0.30]
		"previous": []
	}
	"route": "followup"
	"renderers": ["ca.mcgill.a11y.image.renderer.Text"],
	/* omitted for brevity */
}
```
The server receives and validates the request. In the server, only two services are configured to meet the route requirement:
```yml
services:
  # omitted for brevity
  follow-up-preprocessor:
    image: example-follow-up-preprocessor:latest
    labels:
      ca.mcgill.a11y.image.preprocessor: 1
      ca.mcgill.a11y.image.route: followup
  follow-up-capable-handler:
    image: example-follow-up-handler:
    labels:
      ca.mcgill.a11y.image.handler: enable
      ca.mcgill.a11y.image.route: default,followup
  # omitted for brevity
```
The orchestrator first routes the request to the above preprocessor, which determines that the frisbee is pink with purple polka dots. The handler then responds with a request that includes the text "This frisbee seems to be pink with purple polka dots."

Based on user configuration, the client then shows this information either as an annotation on the tactile SVG, separately on-screen in Braille, or via a speech synthesizer.

**Comment by jeffbl** [comment]
Misc thoughts:
- some clients may not be able to provide text from a spoken followup request for the `query` field. This may include the Monarch with default software. For this grant, text is acceptable, but I can imagine future situations where the STT has to happen on the server using something like whisper.
- `focus` field implies only one rectangular region of focus. Is it worth expanding this to allow the user to outline a region, or specify two distinct regions, e.g. "is the mitochondria here [region] or here [region]?" Is this type of query even possible? @VenissaCarolQuadros can we flesh out some concrete examples of followup queries to make sure that a single rectangle of focus gives us enough flexibility to handle likely cases?
- important to specify that if a client does not handle multiple renderings, it is responsible for choosing which one to render? Or client MUST render first one returned (which would let the server decide what shows up on such clients)

@AndyBaiMQC any comments, especially on the `previous` field?

**Comment by VenissaCarolQuadros** [comment]
Outlining a region might be tricky given the still somewhat limited accuracy of the 'touch' sensor but I can think of a few scenarios where multiple regions might help e.g.
- Tell me the difference between these two parts/ or what happens between these two stages [region] [region] (or any kind of comparison of this sort especially when it's hard to describe what the two regions are called). 
- What major landmarks are located in these parts of the map? [region] [region] [region].... 

but there are probably more. 

I don't think we currently handle queries of the type "is the mitochondria here [region] or here [region]?" i.e. the region is not given context within the query. For the Monarch specifically, I see some technical complications to having this capability since the SpeechRecognizer runs only on the main thread and needs to be stopped repeatedly to allow anything else to happen.

**Comment by JRegimbal** [comment]
@jeffbl responding point by point
* That's fair. It's just to have a limit right now, but I'd be happy to open it up if need be, but I don't want to specify it as a hard requirement off the bat and have to set up ASR on our servers.
* I worry about how this could be encoded and interpreted on the server if it's any number of gestures with arbitrary regions or points...I can see the utility, but if it's too broad I don't know if that will impact our ability to produce anything useful.
* If a client can't handle multiple renderings, then it does whatever it does! The specification only handles the communications between computers and recommends that multiple renderings be supported since they can be sent back. Since the server doesn't know what the client can do, and the order of renderings returned is arbitrary, I think it's best to not promise anything here.

**Comment by jeffbl** [comment]
I have no problem with any of these constraints for now since you are correct that they are not trivial/involve tradeoffs. My guess is that the first one (voice recognition) may be the first to need revision, since I don't know if there is a web STT that would be viable in the extension, and I expect we'll get pressure to do followups for audio experiences at some point. But let's do this first, then revise when we need to deal with new use cases.

**Comment by JRegimbal** [comment]
Assigned to @AndyBaiMQC to check if this is workable from the ML perspective.

**Comment by VenissaCarolQuadros** [comment]
Hi @AndyBaiMQC, do you have any preliminary comments on this?  I'm particularly interested in your input on the 'previous' field— specifically, whether it would work if they are 2-length arrays with both the query and response as strings.

**Comment by AndyBaiMQC** [comment]
@VenissaCarolQuadros @jeffbl Sorry for the late reply. It is completely doable and should be a 1-liner as part of a parsing function. From ML technicality side of things, this design has no impact on how we do prompt engineering. 

Looking at the definition (and credit to @JRegimbal):

`The first element of this array is a string representing the query for a request, and the second element is a string or object containing the corresponding response's data. `

We just need to append the query as `arr[0]` and parse accordingly for `arr[1]`

In summary, take a 'thumb-up' for approval from me on this end.

**Comment by JRegimbal** [comment]
Closing since everyone directly involved has now commented. Please ping with future issues/suggestions.
